sql/parser: produce context-sensitive help during error recovery.

knz opened the pull request:
(forked off #13540) 

This patch leverages the classical error recovery mechanism of LALR
parsers to generate context-sensitive help messages. The machinery to
achieve this is automatically generated from comments in the grammar.

For example:

```
stmt:
...
| backup_stmt     // EXTEND WITH HELP: BACKUP
...

// %Help: BACKUP - back up data to external storage
// %Category: CCL
// %Text:
// BACKUP <targets...> TO <location...>
//        [ AS OF SYSTEM TIME <expr> ]
//        [ INCREMENTAL FROM <location...> ]
//        [ WITH OPTIONS ( <options...> ) ]
//
// Targets:
//    TABLE <pattern> [ , <patterns...> ]
//    DATABASE <databasename> [ , <databasenames...> ]
//
// Location:
//    "[scheme]://[host]/[path to backup]?[parameters]"
//
// Options:
//    INTO_DB
//    SKIP_MISSING_FOREIGN_KEYS
//
// %SeeAlso: RESTORE, https://www.cockroachlabs.com/docs/backup.html
backup_stmt:
  BACKUP targets TO string_or_placeholder opt_as_of_clause opt_incremental opt_with_options
  {
    $$.val = &Backup{...}
  }
| BACKUP error // SHOW HELP: BACKUP
```

The strings `EXTEND WITH HELP` and `SHOW HELP` cause extra code to be generated:

```
stmt:
...
| backup_stmt      %prec VALUES
| backup_stmt      HELPTOKEN %prec UMINUS { return helpWith(sqllex, "BACKUP") }
...

backup_stmt:
  BACKUP targets TO string_or_placeholder ...
  { ... }
| BACKUP error  { return helpWith(sqllex, "BACKUP") }
```

and separately in `help_messages.go`:

```go
var helpMessages = map[string]HelpMessageBody{
 ...
 //line sql.y: 1216
  `BACKUP`: {
    ShortDescription: `back up data to external storage`,
//line sql.y: 1217
    Category: hCCL,
//line sql.y: 1218
    Text: `
BACKUP <targets...> TO <location...>
       [ AS OF SYSTEM TIME <expr> ]
       [ INCREMENTAL FROM <location...> ]
       [ WITH OPTIONS ( <options...> ) ]

Targets:
   TABLE <pattern> [ , <patterns...> ]
   DATABASE <databasename> [ , <databasenames...> ]

Location:
   "[scheme]://[host]/[path to backup]?[parameters]"

Options:
   INTO_DB
   SKIP_MISSING_FOREIGN_KEYS
`,
...
}
```

This causes the parser to produce the custom help text as error when a
syntax error is encountered in the corresponding grammar rule.

Of course, some care must be taken to not produce *only* help texts:
if a client enters invalid statements, this should cause a suitable
error message, including e.g. "`syntax error`" with the detailed
position of the error in the input SQL text.

To achieve this, the `error` mechanism is complemented as follows:

- we introduce a new lexical token HELPTOKEN, which is emitted by the
  scanner when the input contains the syntax `?`. Conveniently, this
  is the syntax appended to the SQL text by the `sql` CLI after the
  previous patch.

- we ensure that HELPTOKEN is not used by the grammar, so that a
  syntax error is guaranteed when it is present in the input.

- we only emit the help text as error if the last token encountered
  was HELPTOKEN. If HELPTOKEN was not present, the regular syntax
  error message is reported instead.

References:
- `README.md` in `pkg/sql/parser`
- https://www.gnu.org/software/bison/manual/html_node/Error-Recovery.html
- http://stackoverflow.com/questions/9796608/error-handling-in-yacc

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/16295)
<!-- Reviewable:end -->

knz commented:
NB: this is a separate PR as requested in https://github.com/cockroachdb/cockroach/pull/13540#issuecomment-305886353 to make reviewing easier.

mjibson commented:
Review status: 0 of 6 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/sql/parser/scan.go, line 160 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16295#-Km34DyLefEBkOV-m_yi:-Km34DyLefEBkOV-m_yj:bdma5jw) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6d0eb144a1b737a46dbb6d8a7c02209c6908438/pkg/sql/parser/scan.go#L160)):*
> ```Go
> func (s *Scanner) Error(e string) {
> 	var buf bytes.Buffer
> 	switch s.lastTok.id {
> ```

Why did this need to change from an if to a switch?

---

*[pkg/sql/parser/sql.y, line 896 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16295#-Km32rDRIqIAq0iBB2y-:-Km32rDRIqIAq0iBB2y0:b-1kd740) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6d0eb144a1b737a46dbb6d8a7c02209c6908438/pkg/sql/parser/sql.y#L896)):*
> ```Yacc
> alter_stmt:
>   ALTER error { return helpWith(sqllex, "ALTER") }
> | alter_table_stmt              %prec VALUES
> ```

What happens without these precedence directives?

---

*[pkg/sql/parser/sql.y, line 3326 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16295#-Km380VnzDAhk_tS6wvY:-Km380VoubCx8_6zyuwV:b-imazwv) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6d0eb144a1b737a46dbb6d8a7c02209c6908438/pkg/sql/parser/sql.y#L3326)):*
> ```Yacc
>     $$.val = &From{Tables: $2.tblExprs(), AsOf: $3.asOfClause()}
>   }
> | FROM error { return helpWith(sqllex, "<SOURCE>")  }
> ```

I don't understand why we need to add `error` to all these if helpWith just converts them to normal errors anyway.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16295)*
<!-- Sent from Reviewable.io -->

knz commented:
I am splitting the grammar refactorings into a separate PR #17033. When that merges I will rebase this PR to focus on the help-related changes. The only relevant commit in this PR is the last one.

---

Review status: 0 of 18 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/sql/parser/scan.go, line 160 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16295#-Km34DyLefEBkOV-m_yi:-Kp7B9fo1mRrIjcvRDCH:bl5xt6q) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6d0eb144a1b737a46dbb6d8a7c02209c6908438/pkg/sql/parser/scan.go#L160)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

Why did this need to change from an if to a switch?
</blockquote></details>

It doesn't need to. Reverted.

---

*[pkg/sql/parser/sql.y, line 896 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16295#-Km32rDRIqIAq0iBB2y-:-Kp7BBO2YFdKoE3hqXzU:b-364au3) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6d0eb144a1b737a46dbb6d8a7c02209c6908438/pkg/sql/parser/sql.y#L896)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

What happens without these precedence directives?
</blockquote></details>

Then we would get shift/reduce errors. This is because HELPTOKEN is also implicitly part of the error rule within the alter_table_stmt syntax definition.

---

*[pkg/sql/parser/sql.y, line 3326 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16295#-Km380VnzDAhk_tS6wvY:-Kp7BP93lONQp54PzmUh:b-h4cseh) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6d0eb144a1b737a46dbb6d8a7c02209c6908438/pkg/sql/parser/sql.y#L3326)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

I don't understand why we need to add `error` to all these if helpWith just converts them to normal errors anyway.
</blockquote></details>

These are two separate / independent mechanisms.
The `error` rules invokes help when a user has entered an invalid syntax, for example here `SELECT * FROM 1+2`. 
The `HELPTOKEN` rules invoke help when a user has entered the help query token (`?`). 

The `error` pseudo-rule is very special in LALR generators (yacc/bison/goyacc etc): it causes a special rule that skips all invalid input tokens automatically until error recovery is possible. For more details about `error` see:
- https://www.gnu.org/software/bison/manual/html_node/Error-Recovery.html
- http://stackoverflow.com/questions/9796608/error-handling-in-yacc

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16295#-:-Kp7B-_Z6tY4q76RwyQs:b-hixl1d)*
<!-- Sent from Reviewable.io -->

knz commented:
I have embedded the help strings inside the grammar, following an idea/recommendation by @benesch. PTAL.

knz commented:
@benesch can you check what I did with the Makefile and see if you find it agreeable?

@jordanlewis can you provide feedback on what I did with the grammar? I can also explain if you're not familiar with the underlying principles. Your feedback also good on the Go code.

@benesh @mjibson can you provide feedback on the overall patch, and whether this combines well with the integration with the railroad syntax diagram generation.

knz commented:
Added a README file to explain everything that's going on.

tamird commented on pkg/sql/parser/help_gen_test.sh:
```diff
@@ -0,0 +1,20 @@
+set -euo pipefail
+
+cat <<EOF
+// Code generated by help_gen_test.sh.
```
@knz: FYI https://github.com/golang/go/issues/13560#issuecomment-277804473

knz replied:
Done.

tamird replied:
heh, still not done in this one.

knz replied:
Well spotted! Fixed.

jordanlewis commented:
Very cool. We talked about it offline, but I would like to point out for discussion that having important directives in comments is confusing for the novice CockroachDB grammar hacker. We should weigh this potential trap against inventing a preprocessor syntax that isn't in comments and confusing people's YACC syntax highlighters.

---

Review status: 0 of 16 files reviewed at latest revision, 16 unresolved discussions, some commit checks failed.

---

*[pkg/sql/parser/help.awk, line 45 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16295#-KqxKfDHwvI_ggP3LUOq:-KqxKfDHwvI_ggP3LUOr:b-vmr5aq) ([raw file](https://github.com/cockroachdb/cockroach/blob/e313deffb014694d63962a003bd875fa2789407f/pkg/sql/parser/help.awk#L45)):*
> ```Awk
>     has_summary = index(helprest, "-")
>     if (has_summary > 0) {
> 	summary = substr(helprest, has_summary+1)
> ```

mixed tabs and spaces

---

*[pkg/sql/parser/help.awk, line 55 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16295#-KqxKm7KtYMuB1m4oPap:-KqxKm7KtYMuB1m4oPaq:b-oxawx8) ([raw file](https://github.com/cockroachdb/cockroach/blob/e313deffb014694d63962a003bd875fa2789407f/pkg/sql/parser/help.awk#L55)):*
> ```Awk
> 	printf "    ShortDescription: `%s`,\n", summary
>     }
>     next
> ```

This code slightly scares me. I can see that it works based on its outputs, but I am worried (without a compelling alternative, to be fair) about its maintainability.

Perhaps some comments above each of the rules would help a bit...

---

*[pkg/sql/parser/Makefile, line 31 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16295#-KqxFFEedXTnEvgyztF7:-KqxFFEedXTnEvgyztF8:b-j30l4x) ([raw file](https://github.com/cockroachdb/cockroach/blob/e313deffb014694d63962a003bd875fa2789407f/pkg/sql/parser/Makefile#L31)):*
> ```Makefile
> # joins all of the lines into a single line with a '|' character to be
> # used as a regexp "or" meta-character. Finally, the last '|' character
> # is striped from the string.
> ```

s/striped/stripped/

---

*[pkg/sql/parser/Makefile, line 32 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16295#-KqxFMUJ9wIzGuPtzy-y:-KqxFMUJ9wIzGuPtzy-z:b-mon5zk) ([raw file](https://github.com/cockroachdb/cockroach/blob/e313deffb014694d63962a003bd875fa2789407f/pkg/sql/parser/Makefile#L32)):*
> ```Makefile
> # used as a regexp "or" meta-character. Finally, the last '|' character
> # is striped from the string.
> # Then translate the original syntax file, with the types determines above being
> ```

s/determines/determined/

---

*[pkg/sql/parser/README.md, line 69 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16295#-KqxGNHvvDWE9pJWPrAi:-KqxGNHvvDWE9pJWPrAj:b-v57ldq) ([raw file](https://github.com/cockroachdb/cockroach/blob/e313deffb014694d63962a003bd875fa2789407f/pkg/sql/parser/README.md#L69)):*
> ```Markdown
> - do not continue parsing (the generated parser executes all actions
>   in a big loop; a `return` will interrupt this loop);
> - stop parsing with an indicator an error has occurred (the `helpWith`
> ```

these bullets don't really match the tense of "What this does is:" - so it's a little hard to read.

I would change everything to be the same tense, like:

- halts parsing  (...)
- returns an error, extended with an error message ...

---

*[pkg/sql/parser/README.md, line 241 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16295#-KqxHLGMXyu3-hZEDk0P:-KqxHLGMXyu3-hZEDk0Q:b-4816s4) ([raw file](https://github.com/cockroachdb/cockroach/blob/e313deffb014694d63962a003bd875fa2789407f/pkg/sql/parser/README.md#L241)):*
> ```Markdown
> `HIGHTOKEN`) however conveniently we can also reuse otherwise valid
> tokens that have known relative priorities. We settled in our case on
> `VALUES` (low priority) and `UMINUS` (high priority).
> ```

very nice explanation of all of the above. :+1: 

---

*[pkg/sql/parser/replace_help_rules.awk, line 8 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16295#-KqxHUprkptrjPmndTKD:-KqxHUprkptrjPmndTKE:b-ug5rg) ([raw file](https://github.com/cockroachdb/cockroach/blob/e313deffb014694d63962a003bd875fa2789407f/pkg/sql/parser/replace_help_rules.awk#L8)):*
> ```Awk
>     rulename = prefix
>     if (index($0, "|") > 0) {
> 	rulename = substr(rulename, index($0, "|")+1)
> ```

mixed tabs/spaces

---

*[pkg/sql/parser/sql.y, line 947 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16295#-KqxI7WC3YlrJG8Qy3o0:-KqxI7WC3YlrJG8Qy3o1:b-qk0p4n) ([raw file](https://github.com/cockroachdb/cockroach/blob/e313deffb014694d63962a003bd875fa2789407f/pkg/sql/parser/sql.y#L947)):*
> ```Yacc
> //   ALTER TABLE ... VALIDATE CONSTRAINT <constraintname>
> //   ALTER TABLE ... SPLIT AT <selectclause>
> //   ALTER TABLE ... TESTING_RELOCATE <selectclause>
> ```

do we want to expose testing_relocate to our users?

---

*[pkg/sql/parser/sql.y, line 1225 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16295#-KqxIW0CapfZ-sXC_JpL:-KqxIW0CapfZ-sXC_JpM:bz6c7iw) ([raw file](https://github.com/cockroachdb/cockroach/blob/e313deffb014694d63962a003bd875fa2789407f/pkg/sql/parser/sql.y#L1225)):*
> ```Yacc
> //
> // Targets:
> //    TABLE <pattern> [ , <patterns...> ]
> ```

I find this syntax confusing - I think it should in general be `<entity> [ , ... ]` OR `<entity [, <entity>`, but pluralizing the entity and including dots seems inconsistent.

I'm okay with keeping this though, just a thought.

---

*[pkg/sql/parser/sql.y, line 1440 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16295#-KqxIzdHIMyu3tOhuxt1:-KqxIzdHIMyu3tOhuxt2:b-jdtdv5) ([raw file](https://github.com/cockroachdb/cockroach/blob/e313deffb014694d63962a003bd875fa2789407f/pkg/sql/parser/sql.y#L1440)):*
> ```Yacc
> // %Help: DROP VIEW - remove a view
> // %Category: DDL
> // %Text: DROP VIEW [IF EXISTS] <tablename> [, ...] [CASCADE | RESTRICT]
> ```

See here for example to be consistent with the above it should be `<tablename> [, <tablenames...>]`. But I much prefer this. I think this style should be propagated everywhere.

---

*[pkg/sql/parser/sql.y, line 1697 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16295#-KqxJDMuEdm-EXPVTK30:-KqxJDMuEdm-EXPVTK31:bvo2rz1) ([raw file](https://github.com/cockroachdb/cockroach/blob/e313deffb014694d63962a003bd875fa2789407f/pkg/sql/parser/sql.y#L1697)):*
> ```Yacc
> //
> // Targets:
> //   DATABASE <databasename> [, <databasename>]...
> ```

dots should come inside the `[]`.

---

*[pkg/sql/parser/sql.y, line 3536 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16295#-KqxKD4k9f1PrXK5euxX:-KqxKD4k9f1PrXK5euxY:b-kinxcn) ([raw file](https://github.com/cockroachdb/cockroach/blob/e313deffb014694d63962a003bd875fa2789407f/pkg/sql/parser/sql.y#L3536)):*
> ```Yacc
> //        [ FROM <source> ]
> //        [ WHERE <expr> ]
> //        [ GROUP BY <expr> [,...] ]
> ```

`[ , ... ]`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16295#-:-KqxLDO42rcF_UjZ2hbM:b-62kwv7)*
<!-- Sent from Reviewable.io -->

knz commented:
Thanks for your review.

Regarding an alternative proposal, I 'm welcoming of any suggestion. If you can invent something that seems to work, I'll take it!

---

Review status: 0 of 16 files reviewed at latest revision, 16 unresolved discussions.

---

*[pkg/sql/parser/help.awk, line 45 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16295#-KqxKfDHwvI_ggP3LUOq:-Kr19ztx3woPrXctXT0g:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e313deffb014694d63962a003bd875fa2789407f/pkg/sql/parser/help.awk#L45)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

mixed tabs and spaces
</blockquote></details>

Done.

---

*[pkg/sql/parser/help.awk, line 55 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16295#-KqxKm7KtYMuB1m4oPap:-Kr1BonIyCoSHs--6dmK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e313deffb014694d63962a003bd875fa2789407f/pkg/sql/parser/help.awk#L55)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

This code slightly scares me. I can see that it works based on its outputs, but I am worried (without a compelling alternative, to be fair) about its maintainability.

Perhaps some comments above each of the rules would help a bit...
</blockquote></details>

Done.

---

*[pkg/sql/parser/Makefile, line 31 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16295#-KqxFFEedXTnEvgyztF7:-Kr1BsZA603_kURbJqtf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e313deffb014694d63962a003bd875fa2789407f/pkg/sql/parser/Makefile#L31)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

s/striped/stripped/
</blockquote></details>

Done.

---

*[pkg/sql/parser/Makefile, line 32 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16295#-KqxFMUJ9wIzGuPtzy-y:-Kr1BunI10XMB0xnfLJv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e313deffb014694d63962a003bd875fa2789407f/pkg/sql/parser/Makefile#L32)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

s/determines/determined/
</blockquote></details>

Done.

---

*[pkg/sql/parser/README.md, line 69 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16295#-KqxGNHvvDWE9pJWPrAi:-Kr1CDBPYNn20GNBKn6j:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e313deffb014694d63962a003bd875fa2789407f/pkg/sql/parser/README.md#L69)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

these bullets don't really match the tense of "What this does is:" - so it's a little hard to read.

I would change everything to be the same tense, like:

- halts parsing  (...)
- returns an error, extended with an error message ...
</blockquote></details>

Done.

---

*[pkg/sql/parser/replace_help_rules.awk, line 8 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16295#-KqxHUprkptrjPmndTKD:-Kr1CF94-nOoiiFcOywq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e313deffb014694d63962a003bd875fa2789407f/pkg/sql/parser/replace_help_rules.awk#L8)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

mixed tabs/spaces
</blockquote></details>

Done.

---

*[pkg/sql/parser/sql.y, line 947 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16295#-KqxI7WC3YlrJG8Qy3o0:-Kr199GwXcFFZoDUkeUH:b-unaoom) ([raw file](https://github.com/cockroachdb/cockroach/blob/e313deffb014694d63962a003bd875fa2789407f/pkg/sql/parser/sql.y#L947)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

do we want to expose testing_relocate to our users?
</blockquote></details>

Hmm maybe not. Fixed.

---

*[pkg/sql/parser/sql.y, line 1225 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16295#-KqxIW0CapfZ-sXC_JpL:-Kr19R1RuGFEhvUVyzCb:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/e313deffb014694d63962a003bd875fa2789407f/pkg/sql/parser/sql.y#L1225)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I find this syntax confusing - I think it should in general be `<entity> [ , ... ]` OR `<entity [, <entity>`, but pluralizing the entity and including dots seems inconsistent.

I'm okay with keeping this though, just a thought.
</blockquote></details>

Fixed.

---

*[pkg/sql/parser/sql.y, line 1440 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16295#-KqxIzdHIMyu3tOhuxt1:-Kr19TAN-mzqpCpeQiIT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e313deffb014694d63962a003bd875fa2789407f/pkg/sql/parser/sql.y#L1440)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

See here for example to be consistent with the above it should be `<tablename> [, <tablenames...>]`. But I much prefer this. I think this style should be propagated everywhere.
</blockquote></details>

Done.

---

*[pkg/sql/parser/sql.y, line 1697 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16295#-KqxJDMuEdm-EXPVTK30:-Kr19o9ULZoehLgtaTJT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e313deffb014694d63962a003bd875fa2789407f/pkg/sql/parser/sql.y#L1697)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

dots should come inside the `[]`.
</blockquote></details>

Done.

---

*[pkg/sql/parser/sql.y, line 3536 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16295#-KqxKD4k9f1PrXK5euxX:-Kr19tLrYylswhOOEhv2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e313deffb014694d63962a003bd875fa2789407f/pkg/sql/parser/sql.y#L3536)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

`[ , ... ]`
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16295#-:-Kr1CJU2kWlOaYdQTBf4:b-zbq0gf)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 16 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[pkg/sql/parser/replace_help_rules.awk, line 8 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16295#-KqxHUprkptrjPmndTKD:-Kr6ZZAGb-o_j4Fq95Cj:b-ulel7o) ([raw file](https://github.com/cockroachdb/cockroach/blob/e313deffb014694d63962a003bd875fa2789407f/pkg/sql/parser/replace_help_rules.awk#L8)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Doesn't seem done in patch 4

---

*[pkg/sql/parser/sql.y, line 947 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16295#-KqxI7WC3YlrJG8Qy3o0:-Kr6_aMMda_br1ppKxi2:b4wf81g) ([raw file](https://github.com/cockroachdb/cockroach/blob/e313deffb014694d63962a003bd875fa2789407f/pkg/sql/parser/sql.y#L947)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Hmm maybe not. Fixed.
</blockquote></details>

You left one of them in!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16295#-:-Kr6_mg2lEbQKnvN6FkK:bnfp4nl)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 0 of 16 files reviewed at latest revision, 7 unresolved discussions, some commit checks broke.

---

*[pkg/sql/parser/replace_help_rules.awk, line 8 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16295#-KqxHUprkptrjPmndTKD:-Kr7XWqbKNpX8w2tB226:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e313deffb014694d63962a003bd875fa2789407f/pkg/sql/parser/replace_help_rules.awk#L8)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Doesn't seem done in patch 4
</blockquote></details>

Done.

---

*[pkg/sql/parser/sql.y, line 947 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16295#-KqxI7WC3YlrJG8Qy3o0:-Kr7Xh0UfvRnPPtBJY1P:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e313deffb014694d63962a003bd875fa2789407f/pkg/sql/parser/sql.y#L947)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

You left one of them in!
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16295)*
<!-- Sent from Reviewable.io -->

knz committed:
sql/parser: produce context-sensitive help during error recovery.

This patch leverages the classical error recovery mechanism of LALR
parsers to generate context-sensitive help messages. The machinery to
achieve this is automatically generated from comments in the grammar.

For example:

```
stmt:
...
| backup_stmt     // EXTEND WITH HELP: BACKUP
...

// %Help: BACKUP - back up data to external storage
// %Category: CCL
// %Text:
// BACKUP <targets...> TO <location...>
//        [ AS OF SYSTEM TIME <expr> ]
//        [ INCREMENTAL FROM <location...> ]
//        [ WITH OPTIONS ( <options...> ) ]
//
// Targets:
//    TABLE <pattern> [ , <patterns...> ]
//    DATABASE <databasename> [ , <databasenames...> ]
//
// Location:
//    "[scheme]://[host]/[path to backup]?[parameters]"
//
// Options:
//    INTO_DB
//    SKIP_MISSING_FOREIGN_KEYS
//
// %SeeAlso: RESTORE, https://www.cockroachlabs.com/docs/backup.html
backup_stmt:
  BACKUP targets TO string_or_placeholder opt_as_of_clause opt_incremental opt_with_options
  {
    $$.val = &Backup{...}
  }
| BACKUP error // SHOW HELP: BACKUP
```

The strings `EXTEND WITH HELP` and `SHOW HELP` cause extra code to be generated:

```
stmt:
...
| backup_stmt      %prec VALUES
| backup_stmt      HELPTOKEN %prec UMINUS { return helpWith(sqllex, "BACKUP") }
...

backup_stmt:
  BACKUP targets TO string_or_placeholder ...
  { ... }
| BACKUP error  { return helpWith(sqllex, "BACKUP") }
```

and separately in `help_messages.go`:

```go
var helpMessages = map[string]HelpMessageBody{
 ...
 //line sql.y: 1216
  `BACKUP`: {
    ShortDescription: `back up data to external storage`,
//line sql.y: 1217
    Category: hCCL,
//line sql.y: 1218
    Text: `
BACKUP <targets...> TO <location...>
       [ AS OF SYSTEM TIME <expr> ]
       [ INCREMENTAL FROM <location...> ]
       [ WITH OPTIONS ( <options...> ) ]

Targets:
   TABLE <pattern> [ , <patterns...> ]
   DATABASE <databasename> [ , <databasenames...> ]

Location:
   "[scheme]://[host]/[path to backup]?[parameters]"

Options:
   INTO_DB
   SKIP_MISSING_FOREIGN_KEYS
`,
...
}
```

This causes the parser to produce the custom help text as error when a
syntax error is encountered in the corresponding grammar rule.

Of course, some care must be taken to not produce *only* help texts:
if a client enters invalid statements, this should cause a suitable
error message, including e.g. "`syntax error`" with the detailed
position of the error in the input SQL text.

To achieve this, the `error` mechanism is complemented as follows:

- we introduce a new lexical token HELPTOKEN, which is emitted by the
  scanner when the input contains the syntax `?`. Conveniently, this
  is the syntax appended to the SQL text by the `sql` CLI after the
  previous patch.

- we ensure that HELPTOKEN is not used by the grammar, so that a
  syntax error is guaranteed when it is present in the input.

- we only emit the help text as error if the last token encountered
  was HELPTOKEN. If HELPTOKEN was not present, the regular syntax
  error message is reported instead.

References:
- `README.md` in `pkg/sql/parser`
- https://www.gnu.org/software/bison/manual/html_node/Error-Recovery.html
- http://stackoverflow.com/questions/9796608/error-handling-in-yacc

knz commented:
TFYR!
